ecpy_pulses/item.py

```diff
@@ -0,0 +1,214 @@
+DEP_TYPE = 'ecpy.pulses.items'
+
+
+class Item(HasPrefAtom):
+    """ Base component of a pulse Sequence
+
+    """
+    #: Identifier for the build dependency collector
+    dep_type = Constant(DEP_TYPE).tag(pref=True)
+
+    #: Index identifying the item inside the sequence.
+    index = Int()
+
+    #: Flag to disable a particular item.
+    enabled = Bool(True).tag(pref=True)
+
+    #: Class of the item to use when rebuilding a sequence.
+    item_id = Unicode().tag(pref=True)
+
+    #: Name of the variable which can be referenced in other items.
+    linkable_vars = List()
+
+    #: Reference to the parent sequence.
+    parent = ForwardTyped(sequence)
+
+    #: Reference to the root sequence.
+    root = Value()
+
+    #: Boolean representing whever this item has a root sequence or not
+    has_root = Bool(False)
+
+    #: Mode defining how the def_1 and def_2 attrs shiould be interpreted.
+    def_mode = Enum('Start/Stop',
+                    'Start/Duration',
+                    'Duration/Stop').tag(pref=True)
+
+    #: String representing the item first element of definition : according
+    #: to the selected mode it evaluated value will either be used for the
+    #: start instant, or duration of the item.
+    def_1 = Unicode().tag(pref=True)
+
+    #: String representing the item second element of definition : according
+    #: to the selected mode it evaluated value will either be used for the
+    #: duration, or stop instant of the item.
+    def_2 = Unicode().tag(pref=True)
+
+    #: Actual start instant of the item with respect to the beginning of the
+    #: root sequence. The unit of this time depends of the setting of the
+    #: context.
+    start = Float()
+
+    #: Actual duration of the item. The unit of this time depends of the
+    #: setting of the context.
+    duration = Float()
+
+    #: Actual stop instant of the item with respect to the beginning of the
+    #: root sequence. The unit of this time depends of the setting of the
+    #: context.
+    stop = Float()
+
+    #: Evaluated entries by the eval_entries method
+    eval_entries = Dict()
+
+    def eval_entries(self, root_vars, sequence_locals, missings, errors):
+        """ Attempt to eval the  def_1 and def_2 parameters of the item.
+
+        Parameters
+        ----------
+        root_vars : dict
+            Dictionary of global variables for the all items. This will
+            tipically contains the i_start/stop/duration and the root vars.
+
+        sequence_locals : dict
+            Dictionary of variables whose scope is limited to this item
+            parent.
+
+        missings : set
+            Set of unfound local variables.
+
+        errors : dict
+            Dict of the errors which happened when performing the evaluation.
+
+        Returns
+        -------
+        flag : bool
+            Boolean indicating whether or not the evaluation succeeded.
+
+        """
+        # Flag indicating good completion.
```

Comment:
Add a TODO about using the cache to avoid recomputation and perhaps open an issue to track it.